src/protocolsupport/protocol/packet/middleimpl/clientbound/play/v_pe/DeclareCommands.java

```diff
@@ -0,0 +1,473 @@
+		private final int nameIndex;
+
+		public PEArgumentNode(CommandNode node, PECommandsStructure peStruct) {
+			this.name = node.getName();
+			this.argType = node.getArgType();
+
+			// Cache literal index, if this is a literal argument
+			if (this.argType == null) {
+				this.nameIndex = peStruct.registerLiteral(this.name);
+			} else {
+				this.nameIndex = -1;
+			}
+		}
+
+		public String getName() {
+			return name;
+		}
+
+		public String getArgType() {
+			return argType;
+		}
+
+		public int getNameIndex() {
+			return nameIndex;
+		}
+	}
+
+	private static class PECommandsStructure {
+		private Map<CommandNode, List<List<PEArgumentNode>>> overloadsRegistry = new HashMap<>();
+		private LinkedHashMap<String, Integer> literalRegistry = new LinkedHashMap<>();
+
+		/**
+		 * Register a list of overloads for a top-level node.
+		 */
+		public void registerOverloads(CommandNode node, List<List<PEArgumentNode>> overloads) {
+			overloadsRegistry.put(node, overloads);
+		}
+
+		public List<List<PEArgumentNode>> getOverloads(CommandNode node) {
+			return overloadsRegistry.get(node);
+		}
+
+		/**
+		 * Register a literal in the internal literal registry if it was new, and return its index value.
+		 */
+		public int registerLiteral(String literal) {
+			return literalRegistry.computeIfAbsent(literal, k -> literalRegistry.size() );
+		}
+
+		/**
+		 * Return an ordered set of all registered literals.
+		 */
+		public Set<String> getLiteralArray() {
+			return literalRegistry.keySet();
+		}
+	}
+
+	public DeclareCommands(ConnectionImpl connection) {
+		super(connection);
+	}
+
+	@Override
+	public void readFromServerData(ByteBuf from) {
+		// In theory, we could read this in the superclass. However, right now only PE needs this data, so save us
+		// the trouble of parsing it for everyone else by doing it here.
+		int length = VarNumberSerializer.readVarInt(from);
+
+		allNodes = new CommandNode[length];
+		for (int i = 0; i < length; i++) {
+			CommandNode node = readCommandNode(from);
+			allNodes[i] = node;
+		}
+
+		rootNodeIndex = VarNumberSerializer.readVarInt(from);
+
+		// For convenience, store all top-level nodes in a separate array. (These correspond to the
+		// actual "commands", i.e. the first literal, like "weather").
+		int[] topLevelNodeIndices = allNodes[rootNodeIndex].getChildren();
+		int i = 0;
+		topLevelNodes = new CommandNode[topLevelNodeIndices.length];
+		for (int index : topLevelNodeIndices) {
+			topLevelNodes[i++] = allNodes[index];
+		}
+	}
+
+	private static CommandNode readCommandNode(ByteBuf from) {
+		byte flags = from.readByte();
+		boolean isPathEnd;
+		int redirect;
+		int[] children = ArraySerializer.readVarIntVarIntArray(from).clone();
+		if ((flags & FLAG_HAS_REDIRECT) != 0) {
+			redirect = VarNumberSerializer.readVarInt(from);
+		} else {
+			redirect = -1;
+		}
+		String name;
+		String argType;
+		String suggestion;
+
+		isPathEnd = ((flags & FLAG_IS_COMMAND_PATH_END) != 0);
```

Comment:
can be actually inlined to constructor call

Reply:
How's that helping anything but changing code for the sake of change?

Reply:
Looks ugly, like if we were writing c. Plus when inspecting code initially it all looks like you have muptiple places where this is assigned, why it actually isn't. Which also requires more time to read and understand.